src/data/matrix/notation.lean

```diff
@@ -130,6 +130,147 @@ cons_val_succ x u 0
   vec_cons x u i = x :=
 by { fin_cases i, refl }
 
+-- The following definitions and `simp` lemmas are to allow any
+-- numeral-indexed element of a vector given with matrix notation to
+-- be extracted by `simp` (even when the numeral is larger than the
+-- number of elements in the vector, which is taken modulo that number
+-- of elements by virtue of the semantics of `bit0` and `bit1` and of
+-- addition on `fin n`).
+
+/-- `vec_alt0 v` gives a vector with the same length as `v`, but with
+only alternate elements (even-numbered, counting cyclically). -/
+def vec_alt0 (v : fin n → α) (k : fin n) : α := v $ bit0 k
+
+/-- `vec_alt1 v` gives a vector with the same length as `v`, but with
+only alternate elements (odd-numbered, counting cyclically). -/
+def vec_alt1 (v : fin (n + 1) → α) (k : fin (n + 1)) : α := v $ bit1 k
+
+@[simp] lemma cons_val_bit0 (x : α) (u : fin m → α) (i : fin (m + 1)) :
+  vec_cons x u (bit0 i) = vec_alt0 (vec_cons x u) i :=
+rfl
+
+@[simp] lemma cons_val_bit1 (x : α) (u : fin m → α) (i : fin (m + 1)) :
+  vec_cons x u (bit1 i) = vec_alt1 (vec_cons x u) i :=
+rfl
+
+/-- `vec_join u v` joins two vectors of lengths `m` and `n` to produce
```

Comment:
I'm slightly surprised this doesn't already exist!

Reply:
See also #4406, in particular https://github.com/leanprover-community/mathlib/pull/4406/files#diff-0ad6543ff7afc0f24fe40302b8a0b440R84

Reply:
Despite the doc string there, I definitely needed the extra argument here to give control over definitional equality for the vector length (otherwise there's a problem comparing a vector of length `m + 1 + n` with one of length `m + n + 1` in `cons_join`).

Reply:
My own experience with `finvec`-like structures tells me that the `append` in this PR is much more practical than the one in #4406. (Applying `eq.rec` on data is like the definition of DTT hell.)